contents/javascript/Javascript-modules.md

```diff
@@ -0,0 +1,220 @@
+
+### ES6 modules
+
+Introduced in 2015, ES6 modules is the official implementation of modules in JavaScript. It introduces 2 new syntax `import` and `export` to use modules in JavaScript.
+
+In the example below, `index.js` needs the function `sumOfVariable` from `anExampleModule.js`. So, `anExampleModule.js` module will need to "export" the function using `export` syntax and `index.js` module will need to "import" that function using `import` syntax.
+
+```js
+// anExampleModule.js
+
+var variableOne = 1;
+var variableTwo = 2;
+
+export sumOfVariable() {
+    return variableOne + variableTwo;
+}
+```
+
+```js
+// index.js
+
+import * as anExampleModule from './anExampleModule.js';
+
+anExampleModule.sumOfVariable(); // 3
+```
+
+Due to the very recent adoption of ES6 modules by browsers and there is still some browsers that do not support it, you might not be able to use ES6 modules right away in those unsupported browsers.
+
+There are 2 workarounds for this issue. You can use [transpiler](https://scotch.io/tutorials/javascript-transpilers-what-they-are-why-we-need-them) such as [Babel](https://babeljs.io/) and [bundler](https://medium.com/@gimenete/how-javascript-bundlers-work-1fc0d0caf2da) such as [Webpack](https://webpack.js.org/) to serve your application to those unsupported browsers, or use other module implementations such as CommonJS or module pattern.
+
+### CommonJS
+
+It is one of the most popular modules implementation before ES6 modules is rolled out because it is used by NodeJS itself and hence a lot of NodeJS applications use it.
+
+Using CommonJS is very similar to using ES6 modules because everything is the same except the syntax. CommonJS uses `module.exports` and `require` syntax.
+
+The following example follows the one in ES6 modules, but implemented with CommonJS syntax:
+
+```js
+// anExampleModule.js
+
+var variableOne = 1;
+var variableTwo = 2;
+
+sumOfVariable = function() {
+    return variableOne + variableTwo;
+}
+
+module.exports = {
+  sumOfVariable: sumOfVariable,
+};
+```
+
+```js
+// index.js
+
+var anExampleModule = require('./anExampleModule.js');
+
+anExampleModule.sumOfVariable(); // 3
+```
+
+While CommonJS and ES6 modules may look alike, ES6 modules has various advantages such as support for asynchronous loading, [tree shaking](https://medium.com/@netxm/what-is-tree-shaking-de7c6be5cadd), static code analysis, etc. (these features will not be covered in this article). But this does not make CommonJS to be obsolete because ES6 modules is not natively supported by NodeJS at the time this article is written.
+
+CommonJS is mainly used in NodeJS applications, but you can also use it outside of NodeJS by using bundler. It does not need transpiler like ES6 modules. Nevertheless, if your project requirement does not allow you to use bundler, there is still a modules implementation that does not need any external tool.
```

Comment:
`there is still a modules implementation that does not need any external tool.` -> `the approach given in the next section does not need any external tool.`